docs/docsite/rst/dev_guide/developing_modules_general.rst

```diff
@@ -1,333 +1,313 @@
-        main()
+- Run the playbook and analyze the output: `$ ansible-playbook ./testmod.yml`
 
-.. note::
-    Prior to Ansible-2.1.0, importing only what you used from
-    :mod:`ansible.module_utils.basic` did not work.  You needed to use
-    a wildcard import like this:
+Debugging (local)
+=================   
 
-.. code-block:: python
+If you want to break into a module and step through with the debugger, locally running the module you can do:
 
-        from ansible.module_utils.basic import *
+- Set a breakpoint in the module: `import pdb; pdb.set_trace()`
+- Run the module on the local machine: `$ python -m pdb ./my_new_test_module.py ./args.json`
 
-And instantiating the module class like:
+Debugging (remote)
+==================
 
-.. code-block:: python
+In the event you want to debug a module that is running on a remote target (i.e. not localhost), one way to do this is the following:
 
-    def main():
-        module = AnsibleModule(
-            argument_spec = dict(
-                state     = dict(default='present', choices=['present', 'absent']),
-                name      = dict(required=True),
-                enabled   = dict(required=True, type='bool'),
-                something = dict(aliases=['whatever'])
-            )
-        )
+- On your controller machine (running Ansible) set `ANSIBLE_KEEP_REMOTE_FILES=1` (this tells Ansible to retain the modules it sends to the remote machine instead of removing them)
+- Run your playbook targetting the remote machine and specify ``-vvvv`` (the verbose output will show you many things, including the remote location that Ansible uses for the modules)
+- Take note of the remote path Ansible used on the remote host
+- SSH into the remote target after the completion of the playbook
+- Navigate to the directory (most likely it is going to be your ansible remote user defined or implied from the playbook: ``~/.ansible/tmp/ansible-tmp-...``)
+- Here you should see the module that you executed from your Ansible controller, but this is the zipped file that Ansible sent to the remote host. You can run this by specifying ``python my_test_module.py`` (not necessary)
+- To debug, though, we will want to extra this zip out to the original module format: ``python my_test_module.py explode`` (Ansible will expand the module into ``./debug-dir``)
+- Navigate to ``./debug-dir`` (notice that unzipping has caused the generation of ``ansible_module_my_test_module.py``)
+- Modify or set a breakpoint in the unzipped module
+- Ensure that the unzipped module is executable: ``$ chmod 755 ansible_module_my_test_module.py``
+- Run the unzipped module directly passing the args file: ``$ ./ansible_module_my_test_module.py args`` (args is the file that contains the params that were originally passed. Good for repro and debugging)
 
-The :class:`AnsibleModule` provides lots of common code for handling returns, parses your arguments
-for you, and allows you to check inputs.
+Unit testing
```

Comment:
Hum, I think this section can be removed and we can just document integration testing


Still need to document the quickstart for adding a test though it's basically

1) Create test playbooks under `test/integration/targets/$modulename`
2) Look at existing tests for examples

For more info see http://docs.ansible.com/ansible/dev_guide/testing_integration.html